lib/Dialect/TritonGPU/Transforms/Combine.cpp

```diff
@@ -575,16 +578,75 @@ static int computeCapabilityToMMAVersion(int computeCapability) {
 static SmallVector<int64_t, 2>
 mmaVersionToShapePerWarp(int version, const ArrayRef<int64_t> &shape,
                          int numWarps) {
-  if (version == 1) {
+  if (version == 1)
     return {16, 16};
-  } else if (version == 2) {
+  else if (version == 2)
     return {16, 8};
-  } else {
+  else {
     assert(false && "version not supported");
     return {0, 0};
   }
 }
 
+inline unsigned clamp(unsigned x, unsigned a, unsigned b) {
+  unsigned lo = std::min(a, b);
+  unsigned hi = std::max(a, b);
+  return std::min(std::max(x, lo), hi);
+}
+
+template <int version>
+SmallVector<unsigned, 2> warpsPerTile(const ArrayRef<int64_t> &shape,
+                                      int numWarps);
+
+template <>
+SmallVector<unsigned, 2> warpsPerTile<1>(const ArrayRef<int64_t> &shape,
+                                         int numWarps) {
+  SmallVector<unsigned, 2> ret = {1, 1};
+  SmallVector<int64_t, 2> shapePerWarp =
+      mmaVersionToShapePerWarp(1, shape, numWarps);
+  bool changed = false;
+  do {
+    changed = false;
+    if (ret[0] * ret[1] < numWarps) {
```

Comment:
Just curious, did you get this from the master branch? If so, can you provide me a link to the original code? Both line 610 and line 614 are checking `ret[0] * ret[1] < numWarps`, I don't understand the logic here but my understanding is shallow.

Reply:
https://github.com/openai/triton/blob/master/lib/codegen/analysis/layout.cc#L227

Reply:
I see, thanks!